### PR TITLE
refactor: differentiate between Schema Field and Struct Field

### PR DESCRIPTION
**Pull request by agavra**
### Description 

We now use a different Field object for `LogicalSchema` and `SqlStruct` because:
- `SqlStruct` should not have a `source`
- LogicalSchema fields are better represented as a list than as a `SqlStruct` because it must be flat (it can't be any arbitrary struct)
- Later, I want to add more information to `Column` for schema (e.g. internal names) for quoted identifiers and this change sets that up

### Review Guide

Looks like a big change, but it's mostly just jitter:

- Look at `LogicalSchema`. I reverted #3315 in and now use a `List<Column>` instead of `SqlStruct`
- Look at `Field` (for a struct) and `Column` (for a schema) - they look very similar but they represent two distinct concepts
- Removed `FieldName`

The rest is fallout.

### Testing done 

- `mvn clean install`

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java [hunk 1]
```diff
@@ -147,20 +148,20 @@ public SqlStruct value() {
...
   /**
-   * Find the index of the field with the supplied exact {@code fullFieldName}.
+   * Find the index of the field with the supplied exact {@code fullColumnName}.
    *
-   * @param fullFieldName the exact name of the field to get the index of.
+   * @param fullColumnName the exact name of the field to get the index of.
    * @return the index if it exists or else {@code empty()}.
    */
-  public OptionalInt valueFieldIndex(final String fullFieldName) {
+  public OptionalInt valueFieldIndex(final String fullColumnName) {
```
Comment (big-andy-coates) [inline ksql-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java]:
```suggestion
  public OptionalInt valueColumnIndex(final String fullColumnName) {
```

I won't continue with calling out every use of `field` - you get the general idea ;) . Search for field in the class...

File: ksql-common/src/main/java/io/confluent/ksql/schema/ksql/types/Field.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  private final String name;
+  private final SqlType type;
+
+  public static Field of(final String name, final SqlType type) {
+    return new Field(name, type);
+  }
+
+  private Field(final String name, final SqlType type) {
+    this.name = Objects.requireNonNull(name, "name");
+    this.type = Objects.requireNonNull(type, "type");
```
Comment (big-andy-coates) [inline ksql-common/src/main/java/io/confluent/ksql/schema/ksql/types/Field.java]: throw on empty / non-trimmed field name?

File: ksql-common/src/main/java/io/confluent/ksql/util/StringUtil.java [hunk 3]
```diff
@@ -15,34 +15,25 @@
...
 package io.confluent.ksql.util;
 
-import java.util.List;
+import io.confluent.ksql.schema.ksql.FormatOptions;
 
 public final class StringUtil {
 
-  private StringUtil() {
-  }
+  private StringUtil() { }
```
Comment (big-andy-coates) [inline ksql-common/src/main/java/io/confluent/ksql/util/StringUtil.java]: The way the file was is inline with our coding style.  This change will just get reverted by anyone who formats the file. 

I'd revert this.

```suggestion
  private StringUtil() {
  }
```

File: ksql-common/src/main/java/io/confluent/ksql/util/StringUtil.java [hunk 4]
```diff
@@ -15,34 +15,25 @@
...
-    int cnt = 0;
-    for (final Object obj : objs) {
-      if (cnt > 0) {
-        sb.append(delimiter);
-      }
-      sb.append(obj);
-      cnt += 1;
-    }
-    return sb.toString();
+  public static String escape(final String string, final FormatOptions formatOptions) {
```
Comment (big-andy-coates) [inline ksql-common/src/main/java/io/confluent/ksql/util/StringUtil.java]: Hummm.... if we're adding `escape` to this class then we should rename it to something more _identifier_ specific.  It's not longer _string_ specific.

I know we already have a `IdentifierUtil` in the parser, so you'll have to come up with something else.  Maybe `Identifiers` or `IdentifierFormatUtil`?

Making the name more specific stops people dumping any old stuff in here.  A more targeted class is good.

File: ksql-common/src/test/java/io/confluent/ksql/schema/ksql/ColumnTest.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    new EqualsTester()
+        .addEqualityGroup(
+            Column.of("someName", SqlTypes.INTEGER),
+            Column.of("someName", SqlTypes.INTEGER)
+        )
+        .addEqualityGroup(
+            Column.of("someName".toUpperCase(), SqlTypes.INTEGER)
+        )
+        .addEqualityGroup(
+            Column.of("different", SqlTypes.INTEGER)
```
Comment (big-andy-coates) [inline ksql-common/src/test/java/io/confluent/ksql/schema/ksql/ColumnTest.java]: nit: superfluous equality group. 

The group above tests case-sensitivity. But it its case sensitive on the name, it's definitely sensitive to different names.

File: ksql-common/src/test/java/io/confluent/ksql/schema/ksql/ColumnTest.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        is("`SomeSource`.`SomeName` INTEGER"));
+  }
+
+  @Test
+  public void shouldToStringWithReservedWords() {
+    // Given:
+    final FormatOptions options = FormatOptions.of(
+        identifier -> identifier.equals("reserved")
+            || identifier.equals("word")
+            || identifier.equals("reserved.name")
```
Comment (big-andy-coates) [inline ksql-common/src/test/java/io/confluent/ksql/schema/ksql/ColumnTest.java]: don't think this line is required.

File: ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java [hunk 7]
```diff
@@ -520,7 +520,7 @@ private void visitSelectStar(final AllColumns allColumns) {
             ? source.getAlias() + "_"
             : "";
 
-        for (final Field field : source.getDataSource().getSchema().fields()) {
+        for (final Column field : source.getDataSource().getSchema().fields()) {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java]:
```suggestion
        for (final Column column : source.getDataSource().getSchema().fields()) {
```

File: ksql-engine/src/main/java/io/confluent/ksql/analyzer/SourceSchemas.java [hunk 8]
```diff
@@ -110,12 +110,12 @@ boolean matchesNonValueField(final String fieldName) {
   }
 
   private static Set<String> nonValueFieldNames(final LogicalSchema schema) {
-    final Set<String> fieldNames = schema.metadata().fields().stream()
-        .map(Field::fullName)
+    final Set<String> fieldNames = schema.metadata().stream()
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/analyzer/SourceSchemas.java]:
```suggestion
    final Set<String> columnNames = schema.metadata().stream()
```

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/CodeGenRunner.java [hunk 9]
```diff
@@ -177,7 +177,7 @@ private Visitor(
       this.expressionTypeManager = expressionTypeManager;
     }
 
-    private void addParameter(final Field schemaField) {
+    private void addParameter(final Column schemaField) {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/codegen/CodeGenRunner.java]:
```suggestion
    private void addParameter(final Column schemaColumn) {
```

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/CodeGenRunner.java [hunk 10]
```diff
@@ -318,7 +318,7 @@ public Object visitQualifiedNameReference(
       return null;
     }
 
-    private Field getRequiredField(final String fieldName) {
+    private Column getRequiredField(final String fieldName) {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/codegen/CodeGenRunner.java]:
```suggestion
    private Column getRequiredColumn(final String columnName) {
```

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java [hunk 11]
```diff
@@ -268,7 +268,7 @@ private String formatExpression(final Expression expression) {
         final Void context
     ) {
       final String fieldName = formatQualifiedName(node.getName());
-      final Field schemaField = schema.findValueField(fieldName)
+      final Column schemaField = schema.findValueField(fieldName)
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]:
```suggestion
      final Column schemaColumn = schema.findValueField(fieldName)
```

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java [hunk 12]
```diff
@@ -282,7 +282,7 @@ private String formatExpression(final Expression expression) {
         final Void context
     ) {
       final String fieldName = node.toString();
-      final Field schemaField = schema.findValueField(fieldName)
+      final Column schemaField = schema.findValueField(fieldName)
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]:
```suggestion
      final Column schemaColumn = schema.findValueField(fieldName)
```

etc, etc...

File: ksql-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java (hunk 1 above)
Comment (agavra) [reply ksql-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java]: 😂

File: ksql-common/src/main/java/io/confluent/ksql/schema/ksql/types/Field.java (hunk 2 above)
Comment (agavra) [reply ksql-common/src/main/java/io/confluent/ksql/schema/ksql/types/Field.java]: Added unit tests for these so it isn't missed again!

File: ksql-common/src/main/java/io/confluent/ksql/util/StringUtil.java (hunk 3 above)
Comment (agavra) [reply ksql-common/src/main/java/io/confluent/ksql/util/StringUtil.java]: 😢 okay

**Pull request merged**